### PR TITLE
8306374: (bf) Improve performance of DirectCharBuffer::append(CharSequence[,int,int])

### PR DESCRIPTION
**Pull request by bplb**
Use `getChars()` to copy characters to an intermediate local `char[]` array and then `putArray()` to copy the chars into the buffer itself.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306374](https://bugs.openjdk.org/browse/JDK-8306374): (bf) Improve performance of DirectCharBuffer::append(CharSequence[,int,int])

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author) ⚠️ Review applies to [52c94582](https://git.openjdk.org/jdk/pull/13522/files/52c94582da492e8f19f949900f0317b2a64eca7c)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13522/head:pull/13522` \
`$ git checkout pull/13522`

Update a local copy of the PR: \
`$ git checkout pull/13522` \
`$ git pull https://git.openjdk.org/jdk.git pull/13522/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13522`

View PR using the GUI difftool: \
`$ git pr show -t 13522`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13522.diff">https://git.openjdk.org/jdk/pull/13522.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13522#issuecomment-1513954908)

### CODE REVIEW

**Comment by bplb** [comment]
Appending a sequence of 32768 characters gave the following results:

**Before**
```
Benchmark                                         Mode  Cnt      Score      Error  Units
CharBufferAppend.appendStringBufferToDirect      thrpt   10  42746.409 ±  436.898  ops/s
CharBufferAppend.appendStringBuilderToDirect     thrpt   10  39007.579 ±  934.130  ops/s
CharBufferAppend.appendStringToDirect            thrpt   10  47820.006 ± 1142.298  ops/s
CharBufferAppend.appendSubStringBufferToDirect   thrpt   10  79981.671 ± 2248.619  ops/s
CharBufferAppend.appendSubStringBuilderToDirect  thrpt   10  82032.943 ±  890.607  ops/s
CharBufferAppend.appendSubStringToDirect         thrpt   10  80458.162 ± 1396.552  ops/s
```
**After**
```
Benchmark                                         Mode  Cnt       Score      Error  Units
CharBufferAppend.appendStringBufferToDirect      thrpt   10   41448.953 ± 1176.265  ops/s
CharBufferAppend.appendStringBuilderToDirect     thrpt   10  317041.262 ± 4242.432  ops/s
CharBufferAppend.appendStringToDirect            thrpt   10   47376.189 ±  479.143  ops/s
CharBufferAppend.appendSubStringBufferToDirect   thrpt   10  551124.908 ± 4773.753  ops/s
CharBufferAppend.appendSubStringBuilderToDirect  thrpt   10  616143.045 ± 9020.091  ops/s
CharBufferAppend.appendSubStringToDirect         thrpt   10  605963.509 ± 6222.007  ops/s
```
Throughput is either effectively unchanged or increased by a factor of 6 to 8.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 1]
```diff
@@ -446,6 +446,67 @@ class Direct$Type$Buffer$RW$$BO$
...
+        } while (start < end);
+
+        position(pos + length);
+
+        return this;
+    }
+
+    public $Type$Buffer append(CharSequence csq) {
+#if[rw] 
+        if (csq instanceof StringBuilder)
```
[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:489]: In the development of #13415 we found that those cases were not improved by the heap buffer override, but I think you are correct in that that change should also be measured before we throw it out.

[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:489]: Changing this to also allow `String` and `StringBuffer` does not improve throughput. The reasons are that both of these use the already optimized case of putting a `String` and their `toString()` methods do not copy.

[the remark this answers is not in the document]
Comment (liach) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:489]: Thanks for the investigation.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 2]
```diff
@@ -446,6 +446,67 @@ class Direct$Type$Buffer$RW$$BO$
...
+        checkSession();
+
+        int length = end - start;
+        int pos = position();
+        int lim = limit();
+        int rem = (pos <= lim) ? lim - pos : 0;
+        if (length > rem)
+            throw new BufferOverflowException();
+
+        char[] buf = new char[APPEND_BUF_SIZE];
```
[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: So changed in [d81313a](https://github.com/openjdk/jdk/pull/13522/commits/d81313a6404a8cf7f39b3ab532f8c6b2bbd272c1).

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 3]
```diff
@@ -452,15 +452,18 @@ class Direct$Type$Buffer$RW$$BO$
         checkSession();
 
         int length = end - start;
+        Objects.checkFromIndexSize(start, length,
+                                   csq == null ? 0 : csq.length());
```
Comment (liach) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:456]:
```suggestion
        Objects.checkFromIndexSize(start, length, csq.length());
```
Aren't the calls to `appendChars` already guarded by `instanceof` checks, which produces `false` for `null`?

Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: Oh, I think you are correct. Removed in 52c9458.

**Review by liach — APPROVED** [review]

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 4]
```diff
@@ -446,6 +446,69 @@ class Direct$Type$Buffer$RW$$BO$
 
     // --- Methods to support CharSequence ---
 
+    private static final int APPEND_BUF_SIZE = 1024;
+
+    private $Type$Buffer appendChars(CharSequence csq, int start, int end) {
+        checkSession();
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:452]: This template is for direct buffers so I'm not 100% sure that checkSession is needed here. If the session is closed then SMA.copyMemory used in putArray will throw.

Comment (ExE-Boss) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:452]: This is probably to avoid allocating the temporary `char[] buf` array when the session is known to be closed.

Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: The primary direct buffer constructor is
```
    // Primary constructor
    //
    Direct$Type$Buffer$RW$(int cap) {                   // package-private
```
so in general the `segment` will be `null` and checking it provides no value.

Comment (bplb) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: > This template is for direct buffers so I'm not 100% sure that checkSession is needed here.

Commit 739c33a removes `checkSession()`.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 5]
```diff
@@ -446,6 +446,67 @@ class Direct$Type$Buffer$RW$$BO$
 
     // --- Methods to support CharSequence ---
 
+    private static final int APPEND_BUF_SIZE = 1024;
+
+    private $Type$Buffer appendChars(CharSequence csq, int start, int end) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:451]: I assume we don't need to generate for read-only char buffers.

**Review by AlanBateman — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**